x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/MultiTermsAggregator.java

```diff
@@ -0,0 +1,538 @@
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.bucket.DeferableBucketAggregator;
+import org.elasticsearch.search.aggregations.bucket.terms.BucketPriorityQueue;
+import org.elasticsearch.search.aggregations.bucket.terms.BytesKeyedBucketOrds;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregator;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+
+/**
+ * Collects the {@code multi_terms} aggregation, which functions like the
+ * {@code terms} aggregation, but supports multiple fields that are treated
+ * as a tuple.
+ */
+class MultiTermsAggregator extends DeferableBucketAggregator {
+
+    protected final List<DocValueFormat> formats;
+    protected final TermsAggregator.BucketCountThresholds bucketCountThresholds;
+    protected final BucketOrder order;
+    protected final Comparator<InternalMultiTerms.Bucket> partiallyBuiltBucketComparator;
+    protected final Set<Aggregator> aggsUsedForSorting = new HashSet<>();
+    protected final SubAggCollectionMode collectMode;
+    private final List<TermValuesSource> values;
+    private final boolean showTermDocCountError;
+    private final boolean needsScore;
+    private final List<InternalMultiTerms.KeyConverter> keyConverters;
+
+    private final BytesKeyedBucketOrds bucketOrds;
+
+    protected MultiTermsAggregator(
+        String name,
+        AggregatorFactories factories,
+        AggregationContext context,
+        Aggregator parent,
+        List<ValuesSourceConfig> configs,
+        List<DocValueFormat> formats,
+        boolean showTermDocCountError,
+        BucketOrder order,
+        SubAggCollectionMode collectMode,
+        TermsAggregator.BucketCountThresholds bucketCountThresholds,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, metadata);
+        this.bucketCountThresholds = bucketCountThresholds;
+        this.order = order;
+        partiallyBuiltBucketComparator = order == null ? null : order.partiallyBuiltBucketComparator(b -> b.bucketOrd, this);
+        this.formats = formats;
+        this.showTermDocCountError = showTermDocCountError;
+        if (subAggsNeedScore() && descendsFromNestedAggregator(parent)) {
+            /**
+             * Force the execution to depth_first because we need to access the score of
+             * nested documents in a sub-aggregation and we are not able to generate this score
+             * while replaying deferred documents.
+             */
+            this.collectMode = SubAggCollectionMode.DEPTH_FIRST;
+        } else {
+            this.collectMode = collectMode;
+        }
+        // TODO: Make a function
```

Comment:
Is this a real TODO or just c/p from terms agg?  E.g. is it a todo for this PR or just generally :)